README
NAME
    Text::Match::FastAlternatives - efficient search for many strings

SYNOPSIS
        use Text::Match::FastAlternatives;

        my $expletives = Text::Match::FastAlternatives->new(@naughty);
        while (my $line = <>) {
            print "Do you email your mother with that keyboard?\n"
                if $expletives->match($line);
        }

DESCRIPTION
    This module allows you to search for any of a list of substrings
    ("keys") in a larger string. It is particularly efficient when the set
    of keys is large.

    This efficiency comes at the cost of some flexibility: the keys may not
    contain any control characters or non-ASCII characters; and it cannot do
    case-insensitive matching. If you want case-insensitivity, you have to
    fold case yourself:

        my $expletives = Text::Match::FastAlternatives->new(
            map { lc } @naughty);
        while (my $line = <>) {
            print "Do you email your mother with that keyboard?\n"
                if $expletives->match(lc $line);
        }

    This module is designed as a drop-in replacement for Perl code of the
    following form:

        my $expletives_regex = join '|', map { quotemeta } @naughty;
        $expletives_regex = qr/$expletives_regex/;
        while (my $line = <>) {
            print "Do you email your mother with that keyboard?\n"
                if $line =~ $expletives_regex;
        }

    Text::Match::FastAlternatives can easily perform this test a hundred
    times faster than the equivalent regex, if you have enough keys. The
    more keys it searches for, the faster it gets compared to the regex.

    Modules like Regexp::Trie can build an optimised version of such a
    regex, designed to take advantage of the niceties of perl's regex
    engine. With a large number of keys, this module will substantially
    outperform even an optimised regex like that. In one real-world
    situation with 339 keys, running on Perl 5.8, Regexp::Trie produced a
    regex that ran 857% faster than the naive regex (according to
    Benchmark), but using Text::Match::FastAlternatives ran 18275% faster
    than the naive regex, or twenty times faster than Regexp::Trie's
    optimised regex.

    The enhancements to the regex engine in Perl 5.10 include algorithms
    similar to those in Text::Match::FastAlternatives. However, even with
    very small sets of keys, Perl has to do extra work to be fully general,
    so Text::Match::FastAlternatives is still faster. The difference is
    greater for larger sets of keys. For one test with only 5 keys,
    Text::Match::FastAlternatives was 21% faster than perl-5.10.0; with 339
    keys (as before), the difference was 111% (that is, slightly over twice
    as fast).

METHODS
    Text::Match::FastAlternatives->new(@keys)
        Constructs a matcher that can efficiently search for all of the
        @keys in parallel. Throws an exception if any of the keys are
        undefined, or if any of them contain any control characters or
        non-ASCII characters.

    $matcher->match($target)
        Returns a boolean value indicating whether the $target string
        contains any of the keys in $matcher.

    $matcher->match_at($target, $pos)
        Returns a boolean value indicating whether the $target string
        contains any of the keys in $matcher at position $pos. Returns false
        (without emitting any warning) if $pos is larger than the length of
        $string.

    $matcher->exact_match($target)
        Returns a boolean value indicating whether the $target string is
        exactly equal to any of the keys in $matcher.

CAVEATS
  Subclassing
    Text::Match::FastAlternatives has a "DESTROY" method implemented in XS.
    If you write a subclass with its own destructor, you will need to invoke
    the base destructor, or you will leak memory.

IMPLEMENTATION
    Text::Match::FastAlternatives manages to be so fast by using a trie
    internally. The time to find a match at a given position in the string
    (or determine that there is no match) is independent of the number of
    keys being sought; worst-case match time is linear in the length of the
    longest key. Since a match must be attempted at each position in the
    target string, total worst-case search time is O(*mn*) where *m* is the
    length of the target string and *n* is the length of the longest key.

    The "match_at" and "exact_match" methods only need to find a match at
    one position, so they have worst-case running time of O(min(*n*, *m*)).

SEE ALSO
    <http://en.wikipedia.org/wiki/Trie>, Regexp::Trie, Regexp::Optimizer,
    Regexp::Assemble, perl5100delta.

AUTHOR
    Aaron Crane <arc@cpan.org>

COPYRIGHT
    Copyright 2006, 2007, 2008 Aaron Crane.

    This library is free software; you can redistribute it and/or modify it
    under the terms of the Artistic License, or (at your option) under the
    terms of the GNU General Public License version 2.



Changes
Revision history for Perl extension Text::Match::FastAlternatives

0.01  Sun Dec 17 15:45:18 2006
	- original version

0.02  Sat Dec 23 14:05:27 2006
	- fixed incorrect README file

0.03  Sat Dec 23 14:22:36 2006
	- support Perl 5.8.4 (which has no Newxz() macro)

0.04  Sat Jan  6 13:24:45 2007
	- support older C compilers (C89 rather than C99)

0.05  Thu Nov  1 11:39:17 2007
	- new match_at and exact_match methods

1.00  Mon May  5 12:14:45 2008
	- declare dependency on Perl 5.6
	- update documentation for Perl 5.10
	- label this release as 1.00; the code has been stable for ages
	- if you're already using 0.04 or 0.05, there's no need to upgrade

1.01  Sun Aug 31 16:05:05 2008
	- more flexible: now fully supports Unicode
	- uses less memory: many nodes 376 bytes smaller on 32-bit systems
	- fractionally slower in some cases, but hard to measure the difference

1.02  Mon Sep  1 10:16:22 2008
	- fix test suite to work in Perl 5.10
	- an aside: on 64-bit CPUs, 1.01/1.02 are probably faster than 1.00
	- no code changes; if 1.01 works for you, there's no need to upgrade

1.03  Tue Sep  2 10:48:16 2008
	- fix test suite to work in Perl 5.6
	- no code changes; if 1.02 works for you, there's no need to upgrade

1.04  Sun Sep 28 13:10:48 2008
	- fix bug in match_at on Unicode strings
	- use XSLoader instead of DynaLoader; lower overhead
	- fix test suite Unicode bugs under Perl 5.6


misc/performance.pl
#! /usr/bin/perl

use strict;
use warnings;

# Currently this performance test produces the following results on my
# hardware:
#
#             Rate     rx rxtrie   tmfa
#   rx     0.607/s     --   -87%   -99%
#   rxtrie  4.62/s   661%     --   -94%
#   tmfa    75.5/s 12346%  1535%     --
#
# This isn't quite as good as I claim in the perldoc.  That's because I'm
# testing with fewer keys (246, not 339), and the regex solutions get slower as
# more keys are added, whereas Text::Match::FastAlternatives runs in time that
# is independent of the number of keys.

use blib;

use Text::Match::FastAlternatives;
use File::Slurp qw<read_file>;
use Benchmark qw<cmpthese>;
use List::MoreUtils qw<uniq>;

chomp(my @robots = read_file('t/data/robots.txt'));
chomp(my @uagents = uniq read_file('t/data/user_agents.txt'));

my $rx = do { my $rx = join '|', map { quotemeta } @robots; qr/$rx/i };
my $tmfa = Text::Match::FastAlternatives->new(map { lc } @robots);
my $trie_rx;

my %tests = (
    rx   => sub { for my $ua (@uagents) { my $matched = $ua =~ $rx } },
    tmfa => sub { for my $ua (@uagents) { my $matched = $tmfa->match(lc $ua) } },
);

if (eval { require Regexp::Trie }) {
    $tests{rxtrie} = sub {
        for my $ua (@uagents) { my $matched = lc $ua =~ $trie_rx }
    };
    my $rt = Regexp::Trie->new;
    $rt->add(lc $_) for @robots;
    $trie_rx = $rt->regexp;
}

cmpthese(-60, \%tests);


lib/Text/Match/FastAlternatives.pm
package Text::Match::FastAlternatives;

use strict;
use warnings;

our $VERSION = '1.04';

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

1;

__END__

=head1 NAME

Text::Match::FastAlternatives - efficient search for many strings

=head1 SYNOPSIS

    use Text::Match::FastAlternatives;

    my $expletives = Text::Match::FastAlternatives->new(@naughty);
    while (my $line = <>) {
        print "Do you email your mother with that keyboard?\n"
            if $expletives->match($line);
    }

=head1 DESCRIPTION

This module allows you to search for any of a list of substrings ("keys") in a
larger string.  It is particularly efficient when the set of keys is large.

This efficiency comes at the cost of some flexibility: if you want
case-insensitive matching, you have to fold case yourself:

    my $expletives = Text::Match::FastAlternatives->new(
        map { lc } @naughty);
    while (my $line = <>) {
        print "Do you email your mother with that keyboard?\n"
            if $expletives->match(lc $line);
    }

This module is designed as a drop-in replacement for Perl code of the following
form:

    my $expletives_regex = join '|', map { quotemeta } @naughty;
    $expletives_regex = qr/$expletives_regex/;
    while (my $line = <>) {
        print "Do you email your mother with that keyboard?\n"
            if $line =~ $expletives_regex;
    }

Text::Match::FastAlternatives can easily perform this test a hundred times
faster than the equivalent regex, if you have enough keys.  The more keys it
searches for, the faster it gets compared to the regex.

Modules like Regexp::Trie can build an optimised version of such a regex,
designed to take advantage of the niceties of perl's regex engine.  With a
large number of keys, this module will substantially outperform even an
optimised regex like that.  In one real-world situation with 339 keys,
running on Perl 5.8, Regexp::Trie produced a regex that ran 857% faster than
the naive regex (according to L<Benchmark>), but using
Text::Match::FastAlternatives ran 18275% faster than the naive regex, or
twenty times faster than Regexp::Trie's optimised regex.

The enhancements to the regex engine in Perl 5.10 include algorithms similar
to those in Text::Match::FastAlternatives.  However, even with very small
sets of keys, Perl has to do extra work to be fully general, so
Text::Match::FastAlternatives is still faster.  The difference is greater
for larger sets of keys.  For one test with only 5 keys,
Text::Match::FastAlternatives was 21% faster than perl-5.10.0; with 339 keys
(as before), the difference was 111% (that is, slightly over twice as fast).

=head1 METHODS

=over 4

=item Text::Match::FastAlternatives->new(@keys)

Constructs a matcher that can efficiently search for all of the @keys in
parallel.  Throws an exception if any of the keys are undefined.

=item $matcher->match($target)

Returns a boolean value indicating whether the $target string contains any of
the keys in $matcher.

=item $matcher->match_at($target, $pos)

Returns a boolean value indicating whether the $target string contains any
of the keys in $matcher at position $pos.  Returns false (without emitting
any warning) if $pos is larger than the length of $string.

=item $matcher->exact_match($target)

Returns a boolean value indicating whether the $target string is exactly
equal to any of the keys in $matcher.

=back

=head1 CAVEATS

=head2 Subclassing

Text::Match::FastAlternatives has a C<DESTROY> method implemented in XS.  If
you write a subclass with its own destructor, you will need to invoke the base
destructor, or you will leak memory.

=head2 Interaction with Perl internals

Text::Match::FastAlternatives may change the Perl-internal encoding of
strings passed to C<new> or to its C<match> methods.  This is not considered
a bug, as the Perl-internal encoding of a string is not normally of interest
to Perl code (as opposed to Perl internals).  However, you may encounter
situations where preserving a string's existing encoding is important
(perhaps to work around a bug in some other module).  If so, you may need to
copy scalar variables before matching them:

    $matches++ if $tmfa->match(my $temporary_copy = $original);

=head1 IMPLEMENTATION

Text::Match::FastAlternatives manages to be so fast by using a trie internally.
The time to find a match at a given position in the string (or determine that
there is no match) is independent of the number of keys being sought;
worst-case match time is linear in the length of the longest key.  Since a
match must be attempted at each position in the target string, total worst-case
search time is O(I<mn>) where I<m> is the length of the target string and I<n>
is the length of the longest key.

The C<match_at> and C<exact_match> methods only need to find a match at one
position, so they have worst-case running time of O(min(I<n>, I<m>)).

=head1 SEE ALSO

L<http://en.wikipedia.org/wiki/Trie>, L<Regexp::Trie>, L<Regexp::Optimizer>,
L<Regexp::Assemble>, L<perl5100delta>, L<perlunitut>, L<perlunifaq>.

=head1 AUTHOR

Aaron Crane E<lt>arc@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2006, 2007, 2008 Aaron Crane.

This library is free software; you can redistribute it and/or modify it under
the terms of the Artistic License, or (at your option) under the terms of the
GNU General Public License version 2.

=cut
